spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/ManagedCommitSuite.scala

```diff
@@ -116,6 +123,405 @@ class ManagedCommitSuite
+        // Then we recursively update snapshot to v4 and see no change in commit owner and exit
+        // recursion.
+        DeltaLog.clearCache()
+        resetMetrics()
+        assert(
+          sql(s"SELECT * FROM delta.`$tablePath`").collect().toSet === (0 to 4).map(Row(_)).toSet)
+        assert((cs1.numGetCommitsCalled, cs2.numGetCommitsCalled) === (1, 2))
+        assert((builder1.numBuildCalled, builder2.numBuildCalled) === (1, 2))
+      }
+    }
+  }
+
+  // This test has the following setup:
+  // Setup: There are two commit-stores: cs1 and cs2.
+  //  Any commit recorded in CS1 is also recorded in CS2.
+  //  Any commit recorded in CS2 is not recorded in CS1.
+  // 1. Table is created with CS1 as commit-owner. So 0th commit is persisted in both CS1 and CS2.
+  // 2. Do another commit (v1) on table. Again this will be recorded in both CS1 and CS2.
+  // 3. Take a reference to current DeltaLog and clear the cache. This deltaLog object currently
+  //    points to the latest table snapshot i.e. v1.
+  // 4. Do commit v2 on the table.  Again this will be recorded in both CS1 and CS2.
+  // 5. Do commit v3 on table. As part of this, change commit-owner to CS2. This will be recorded
+  //    in both commit stores CS1, CS2 as CS1 was managing the table when the commit happened.
+  // 6. Do commit v4 on table. This will be recorded in CS2 only as CS2 is the owner now.
+  // 7. Invoke deltaLog.update() on the old deltaLog object which is still pointing to v1.
+  //    - While doing this, we will inject failure in CS2 so that it fails twice when cs2.getCommits
+  //      is called.
+  //    - Because of this old delta log will firstly contact cs1 and get newer commits  i.e. v2/v3
+  //      and create a snapshot out of it. Then it will contact cs2 and fail. So deltaLog.update()
+  //      won't succeed and throw exception. But underlying DeltaLog object now has reference to v3.
+  //      The recorded timestamp for this must be commit timestamp and not clock timestamp.
+  // 8. Invoke deltaLog.update() two more times. 3rd attempt will succeed.
+  //    - the recorded timestamp for this should be clock timestamp.
+  test("correct timestamp is added in CapturedSnapshot") {
+    val batchSize = 10
+    val underlyingCommitStore2 = new InMemoryCommitStore(batchSize)
+    val underlyingCommitStore1 = new InMemoryCommitStore(batchSize) {
+      override def commitImpl(
+          logStore: LogStore,
+          hadoopConf: Configuration,
+          logPath: Path,
+          commitVersion: Long,
+          commitFile: FileStatus,
+          commitTimestamp: Long): CommitResponse = {
+        super.commitImpl(logStore, hadoopConf, logPath, commitVersion, commitFile, commitTimestamp)
+        underlyingCommitStore2
+          .commitImpl(logStore, hadoopConf, logPath, commitVersion, commitFile, commitTimestamp)
+      }
+
+      override protected[delta] def registerBackfill(
```

Comment:
nit: we could simplify the override since this class is anyway scoped to this specific test case?
```suggestion
      override def registerBackfill(
```